django/forms/models.py

```diff
@@ -429,8 +429,17 @@ def _post_clean(self):
                     construct_instance_exclude.append(name)
                 exclude.append(name)
 
-        # Update the model instance with self.cleaned_data.
-        self.instance = construct_instance(self, self.instance, opts.fields, construct_instance_exclude)
+        try:
+            # Update the model instance with self.cleaned_data.
+            self.instance = construct_instance(self, self.instance, opts.fields, construct_instance_exclude)
+        except ValidationError as e:
+            # Ticket #24706
+            # This is necessary to avoid getting
+            # AttributeError: 'ValidationError' object has no attribute 'error_dict'
+            # when trying to _update_errors
+            errors = {}
+            errors = e.update_error_dict(errors)
+            self._update_errors(ValidationError(errors))
```

Comment:
Note for reviewers: I'd be happy to see this machination changed, as I'm not thrilled about the twiddling I'm having to do, but otherwise the `test_setattr_raises_validation_error_non_field` fails hard because `_update_errors` makes certain assumptions, and this was the simplest solution I came up with.